### DISCUSSION TITLE
NVMe SSD driver with PCIe transport available

### DISCUSSION THREAD

**Discussion by rxtxs** [rsta2/circle#494] [Development]
I just discovered Circle and I'm really excited to use it for a high-speed DAQ application on the RPi 5!

I'm planning to use an RPi 5 NVME Hat for fast data storage, but I'm not sure how to access the PCIe bus using the foundations provided by Circle.

Any guidance to point me in the right direction to help get me started is most appreciated.

Thank you in advance.

**Comment by rsta2** [comment]
Unfortunately NVMe is not supported yet. This would require to extend the 
CBcmPCIeHostBridge driver and a new driver for NVMe. At the moment I cannot 
say, when this will be available.

**Reply by rxtxs** [reply]
Thanks for getting back to me so quickly.

It's been awhile since I've done this type of development, but your guidance is enough to get me started. If I make any significant progress on my own, I will reach back to this group.

Thanks again.

**Reply by rsta2** [reply]
It would be great to have this in Circle.

**Comment by rsta2** [comment]
Now after about a year there is a NVMe SSD driver in Circle in *addon/nvme/* on the *develop* branch with a test program in *addon/nvme/sample/*. I tested it with the 256GB Raspberry Pi SSD Kit. Of course it only works on the Raspberry Pi 5. It's in an early stage, so you have to be very careful. DON'T USE IT, if you have data on your SSD, which are important for you! The Kit contained a Samsung SSD with NVMe v1.4 version. v1.3 may also work, but I haven't tested this. Please read the [README](https://github.com/rsta2/circle/blob/develop/addon/nvme/sample/README) in the sample's directory for more info!

The difference in the transfer rate is immense, compared to a SD card. I reached ~280 MBytes/sec with linear read of 16KB blocks and ~30 MBytes/sec with writing on a FAT file system using Chan's FatFs.

Feedback is welcome.

**Comment by rsta2** [comment]
This feature is in Circle 50.1 now.
